src/relations/hacluster.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""hacluster relation hooks and helpers."""
+
+import json
+import logging
+from hashlib import shake_128
+from ipaddress import IPv4Address, ip_address
+from typing import Optional, Union
+
+import ops
+
+HACLUSTER_RELATION_NAME = "ha"
+
+logger = logging.getLogger(__name__)
+
+
+class HACluster(ops.Object):
+    """Defines hacluster functunality."""
+
+    def __init__(self, charm: ops.CharmBase):
+        super().__init__(charm, HACLUSTER_RELATION_NAME)
+
+        self.charm = charm
+
+        self.framework.observe(
+            charm.on[HACLUSTER_RELATION_NAME].relation_changed, self._on_changed
+        )
+        self.framework.observe(charm.on.config_changed, self._on_changed)
+
+    @property
+    def relation(self) -> ops.Relation:
+        """Returns the relations in this model, or None if hacluster is not initialised."""
+        return self.charm.model.get_relation(HACLUSTER_RELATION_NAME)
+
+    def _is_clustered(self) -> bool:
+        for key, value in self.relation.data.items():
+            if isinstance(key, ops.Unit) and key != self.charm.unit:
+                if value.get("clustered") in ("yes", "true"):
+                    return True
+                break
+        return False
+
+    def _on_changed(self, event: Union[ops.RelationChangedEvent, ops.RelationBrokenEvent]) -> None:
+        self.set_vip(self.charm.config.get("vip"))
+
+    def get_unit_juju_status(self) -> ops.StatusBase:
+        """Returns the status of the hacluster if relation exists."""
+        if self.relation and not self.charm.is_externally_accessible(event=None):
+            return ops.BlockedStatus("ha integration used without data-integrator")
+
+        vip = self.charm.config.get("vip")
+        if self.relation and not vip:
+            return ops.BlockedStatus("ha integration used without vip configuration")
+
+        if vip and not self.charm.is_externally_accessible(event=None):
+            return ops.BlockedStatus("vip configuration without data-integrator")
+
+        if self.charm.is_workload_authenticated and self.charm.unit.is_leader() and vip:
+            return ops.ActiveStatus(f"VIP: {vip}")
```

Comment:
How's status messaging is being handled overall on router? Should just be blank when active, given vip can be queried from config?

Reply:
we are setting the VIP in the leader's status message to stay consistent with the pgbouncer implementation

Reply:
will follow up with @dragomirp about the error message, but still leaning towards keeping as is for the moment as pgbouncer has been released to prod with this behavior and we dont want a different experience